src/java/io/bazel/rulesscala/scalac/ScalaCInvoker.java

```diff
@@ -0,0 +1,329 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package io.bazel.rulesscala.scalac;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.devtools.build.lib.worker.WorkerProtocol.WorkRequest;
+import com.google.devtools.build.lib.worker.WorkerProtocol.WorkResponse;
+import io.bazel.rulesscala.jar.JarCreator;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.InputStream;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.lang.reflect.Field;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.nio.file.Files;
+import java.nio.file.FileSystems;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.SimpleFileVisitor;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Enumeration;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.jar.JarOutputStream;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.Map;
+import java.util.TreeMap;
+import scala.Console$;
+import scala.tools.nsc.*;
+import scala.tools.nsc.reporters.ConsoleReporter;
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+/**
+ * This is our entry point to producing a scala target
+ * this can act as one of Bazel's persistant workers.
+ */
+public class ScalaCInvoker {
+  // Mostly lifted from bazel
+  private static void runPersistentWorker() throws IOException {
+    PrintStream originalStdOut = System.out;
+    PrintStream originalStdErr = System.err;
+
+    while (true) {
+      try {
+        WorkRequest request = WorkRequest.parseDelimitedFrom(System.in);
+        if (request == null) {
+          break;
+        }
+
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        int exitCode = 0;
+
+        try (PrintStream ps = new PrintStream(baos)) {
+          System.setOut(ps);
+          System.setErr(ps);
+          Console$.MODULE$.setErrDirect(ps);
+          Console$.MODULE$.setOutDirect(ps);
+          try {
+            processRequest(request.getArgumentsList());
+          } catch (Exception e) {
+            e.printStackTrace();
+            exitCode = 1;
+          }
+        } finally {
+          System.setOut(originalStdOut);
+          System.setErr(originalStdErr);
+        }
+
+        WorkResponse.newBuilder()
+            .setOutput(baos.toString())
```

Comment:
I think we want `new String(baos.toByteArray(), "UTF-8")` here, just to be sure about character sets.